sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -3,45 +3,84 @@
 # ---------------------------------------------------------
 import hashlib
 import logging
+import os.path
+import tempfile
+import threading
 from collections import defaultdict
-from typing import Union, List, Dict
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Dict, Optional
 
+from azure.ai.ml._utils._asset_utils import get_object_hash
+from azure.ai.ml._utils.utils import is_on_disk_cache_enabled
 from azure.ai.ml.constants._common import AzureMLResourceType
 from azure.ai.ml.entities import Component
 from azure.ai.ml.entities._builders import BaseNode
 
 
 logger = logging.getLogger(__name__)
 
+_ANONYMOUS_HASH_PREFIX = "anonymous-component-"
+_YAML_SOURCE_PREFIX = "yaml-source-"
+_CODE_INVOLVED_PREFIX = "code-involved-"
+
+
+@dataclass
+class _CacheContent:
+    component_ref: Component
+    # in-memory hash assume that the code folders are not changed during the run and
+    # use the hash of code path instead of code content to simplify the calculation
+    in_memory_hash: str
+    # on-disk hash will be calculated base on code content if applicable,
+    # so it will work even if the code folders are changed among runs
+    on_disk_hash: Optional[str] = None
+    arm_id: Optional[str] = None
+
 
 class CachedNodeResolver(object):
     """Class to resolve component in nodes with cached component resolution results.
 
     This class is thread-safe if:
-    1) self._resolve_nodes and self._register_node_to_resolve are not called concurrently in the same thread
+    1) self._resolve_nodes is not called concurrently. We guarantee this with a lock in self.resolve_nodes.
     2) self._resolve_component is only called concurrently on independent components
-        a) we have used a component hash to deduplicate components to resolve;
+        a) we have used an in-memory component hash to deduplicate components to resolve first;
         b) nodes are registered & resolved layer by layer, so all child components are already resolved
           when we register a pipeline node;
-        c) potential shared dependencies like compute and data have been resolved before calling self._resolve_nodes.
+        c) potential shared dependencies like compute and data have been resolved before
```

Comment:
All dependencies except code will be resolved before calling self._resolve_component?

Reply:
With current implementation, all dependencies except code and `environment`.

But the statement is not correct. resolution of shared dependencies is thread-safe given they don't involve hash calculation.